models/CVaRmodel.py

```diff
@@ -36,102 +37,78 @@ def rebalancing_model(mu, scenarios, cvar_targets, cvar_alpha, cash, x_old, tran
     float
         Asset weights in an optimal portfolio 
     """ 
-    # define index
+    # Define index
     i_idx = scenarios.columns
-    j_idx = scenarios.index
+    N = i_idx.size
     
-    # number of scenarios
-    scenario_n = scenarios.shape[0]    
-    # variable transaction costs
+    # Number of scenarios
+    T = scenarios.shape[0]    
+    # Variable transaction costs
     c = trans_cost
+
+    # Define variables
+    # - portfolio
+    x = cp.Variable(N, name="x", nonneg=True)
+    # - |x - x_old|
+    absdiff = cp.Variable(N, name="absdiff", nonneg=True)    
+    # - cost
+    cost = cp.Variable(name="cost", nonneg=True) 
+    # - loss deviation
+    vardev = cp.Variable(T, name="vardev", nonneg=True)        
+    # - VaR and CVaR
+    var = cp.Variable(name="var", nonneg=True)
+    cvar = cp.Variable(name="cvar", nonneg=True)
+
+    # Define objective (max expected portfolio return)
+    objective = cp.Maximize(mu.to_numpy() @ x)
+
+    # Define constraints
+    constraints = [
+        # - VaR deviation
+        -scenarios.to_numpy() @ x - var <= vardev,
+
+        # - CVaR limit
+        var + 1/(T * cvar_alpha) * cp.sum(vardev) == cvar,
+        cvar <= cvar_targets,
+
+        # - Cost of rebalancing
+        c * cp.sum(absdiff) == cost,
+        x - x_old <= absdiff,
+        x - x_old >= -absdiff,
+
+        # - Budget
+        x_old.sum() + cash - cp.sum(x) - cost == 0,
+
+        # - Concentration limits
+        x <= max_weight * cp.sum(x)
+    ]
+
+    # Define model
+    model = cp.Problem(objective=objective, constraints=constraints)
+
+    # Solve
+    model.solve(solver=cp.MOSEK, verbose=False)
```

Comment:
Amazing! Will we need some license here? Can we modify the code such that the use can decide if to use paid or non-paid version?